### DISCUSSION TITLE
Try Raspberry Pi Zero W as GS with Release 23.01 image - fpv_video failed

### DISCUSSION THREAD

**Discussion by lida2003** [svpcom/wfb-ng#291] [Questions]
@svpcom 

Hi, I got an idea about using Raspberry Pi Zero W as GS (using it's GPU I hope), so my configuration is as follows:

Raspberry Pi 3B+ (AirUnit)  ==> Raspberry Pi Zero W(GS)

All seems fine, except no video output on HDMI screen.

### Method 1
```
pi@raspberrypi:~ $ sudo systemctl start wifibroadcast@gs
pi@raspberrypi:~ $ sudo systemctl start fpv-video

.... ....
pi@raspberrypi:~ $ sudo systemctl status wifibroadcast@gs
● wifibroadcast@gs.service - WFB profile gs
     Loaded: loaded (/lib/systemd/system/wifibroadcast@.service; disabled; vendor preset: enabled)
     Active: active (running) since Tue 2023-03-14 01:37:57 GMT; 14min ago
   Main PID: 768 (wfb-server)
      Tasks: 6 (limit: 724)
        CPU: 3min 19.882s
     CGroup: /system.slice/system-wifibroadcast.slice/wifibroadcast@gs.service
             ├─768 /usr/bin/python3 /usr/bin/wfb-server gs wfb-rf1
             ├─781 /usr/bin/wfb_rx -p 16 -u 14600 -K /etc/gs.key -i 7669206 wfb-rf1
             ├─782 /usr/bin/wfb_tx -p 144 -u 14601 -K /etc/gs.key -B 20 -G long -S 1 -L 1 -M 1 -k 1 -n 2 -T 0 -i 7669206 wfb-rf1
             ├─783 /usr/bin/wfb_rx -p 0 -c 127.0.0.1 -u 5600 -K /etc/gs.key -i 7669206 wfb-rf1
             ├─790 /usr/bin/wfb_rx -p 160 -u 14800 -K /etc/gs.key -i 7669206 wfb-rf1
             └─791 /usr/bin/wfb_tx -p 32 -u 14801 -K /etc/gs.key -B 20 -G long -S 1 -L 1 -M 1 -k 1 -n 2 -T 0 -i 7669206 wfb-rf1

Mar 14 01:38:18 raspberrypi wfb-server[768]: [server.init_tunnel #info] Tunnel RX: /usr/bin/wfb_rx -p 160 -u 14800 -K /etc/gs.key -i 7669206 wfb-rf1
Mar 14 01:38:18 raspberrypi wfb-server[768]: [server.init_tunnel #info] Tunnel TX: /usr/bin/wfb_tx -p 32 -u 14801 -K /etc/gs.key -B 20 -G long -S 1 -L 1 -M 1 -k 1 -n 2>
Mar 14 01:38:18 raspberrypi wfb-server[768]: [tcp.Port.startListening #info] AntennaFactory starting on 8003
Mar 14 01:38:18 raspberrypi wfb-server[768]: [server.RXProtocol.connectionMade #info] Started tunnel rx
Mar 14 01:38:18 raspberrypi wfb-server[768]: [server.TXProtocol.connectionMade #info] Started tunnel tx
Mar 14 01:38:18 raspberrypi wfb-server[768]: [server.DbgProtocol.lineReceived #info] telem tx: Listen on 14601 for wfb-rf1
Mar 14 01:38:18 raspberrypi wfb-server[768]: [server.DbgProtocol.lineReceived #info] tunnel tx: Listen on 14801 for wfb-rf1
Mar 14 01:38:56 raspberrypi wfb-server[768]: [server.AntennaProtocol.lineReceived #info] New session detected [tunnel rx]: FEC=VDM_RS K=1, N=2, epoch=0
Mar 14 01:39:23 raspberrypi wfb-server[768]: [server.AntennaProtocol.lineReceived #info] New session detected [video rx]: FEC=VDM_RS K=8, N=12, epoch=0
Mar 14 01:42:06 raspberrypi wfb-server[768]: [server.DbgProtocol.lineReceived #info] video rx: 256 packets lost

pi@raspberrypi:~ $ sudo systemctl status fpv-video
● fpv-video.service - FPV video
     Loaded: loaded (/etc/systemd/system/fpv-video.service; disabled; vendor preset: enabled)
     Active: activating (auto-restart) since Tue 2023-03-14 01:41:10 GMT; 1s ago
    Process: 1026 ExecStart=/bin/sh -c gst-launch-1.0 rtspsrc latency=0 location=rtsp://127.0.0.1:8554/wfb ! rtph264depay ! 'video/x-h264,s>
   Main PID: 1026 (code=exited, status=0/SUCCESS)
        CPU: 997ms

Mar 14 01:41:12 raspberrypi systemd[1]: fpv-video.service: Scheduled restart job, restart counter is at 16.
Mar 14 01:41:12 raspberrypi systemd[1]: Stopped FPV video.
```

### Method 2

```
pi@raspberrypi:~ $ gst-launch-1.0 udpsrc port=5600 caps='application/x-rtp, media=(string)video, clock-rate=(int)90000, encoding-name=(string)H264' \
               ! rtph264depay ! avdec_h264 ! clockoverlay valignment=bottom ! autovideosink fps-update-interval=1000 sync=false
WARNING: erroneous pipeline: no element "avdec_h264"
```

----------------
![图片](https://user-images.githubusercontent.com/1296171/224870874-d93c1945-e069-4889-adb7-771d340cd30a.png)

![图片](https://user-images.githubusercontent.com/1296171/224870820-a8c58b9f-5186-4099-9166-aba6bb543dc9.png)

**Comment by lida2003** [comment]
It seems OK now :)

Mate desktop + 720P @ 60FPS , excellent software, @svpcom !

![图片](https://user-images.githubusercontent.com/1296171/225794721-c433f465-543e-47e8-b865-5670f5934ce3.png)

**Reply by lida2003** [reply]
Here is my test result (video + latency numbers FYI)

Release 23.01 need [desktop support](https://blog.csdn.net/lida2003/article/details/129559105) for gs.

- [FPV Camera(RPI 3B+/Zero W+V2.1) | wfb_ng Release 23.01 | H264 decoder](https://blog.csdn.net/lida2003/article/details/129623814)
- [FPV Camera(RPI3+V2.1) | wfb_ng Release 23.01 | Ubuntu 20.04 xfce sw decoder](https://blog.csdn.net/lida2003/article/details/129491517)
- [FPV Camera(RPI3+V2.1) | wfb_ng Release 23.01 | ubuntu20.04 gnome sw decoder](https://blog.csdn.net/lida2003/article/details/129478119)

**Reply by svpcom** [reply]
No, on RPI wfb-ng can show video via mmal (hw accelerated) pipeline. See /etc/systemd/system/fpv-video.service

**Reply by svpcom** [reply]
Run `sudo journalctl -xu fpv-video.service` to see why it failed

**Reply by svpcom** [reply]
Also you need to start rtsp service

**Reply by lida2003** [reply]
> No, on RPI wfb-ng can show video via mmal (hw accelerated) pipeline. See /etc/systemd/system/fpv-video.service

ok, i'll try later.

**Reply by lida2003** [reply]
> No, on RPI wfb-ng can show video via mmal (hw accelerated) pipeline. See /etc/systemd/system/fpv-video.service

**I think you are right. There are a few things happend， which I summarized and promote a few questions below.**

**a) I was using rtsp commands, which is default from the image. And it failed as the title suggest.**
**b) And I try to use following commands in RPI, which uses CPU for H264 decoding. It need desktop support.**

```
gst-launch-1.0 udpsrc port=5600 caps='application/x-rtp, media=(string)video, clock-rate=(int)90000, encoding-name=(string)H264' \
               ! rtph264depay ! avdec_h264 ! clockoverlay valignment=bottom ! autovideosink fps-update-interval=1000 sync=false
```

**c) Which one is better for RPI, RTP legacy or RTP MMAL?**


### RTP legacy
`ExecStart=/bin/sh -c "gst-launch-1.0 udpsrc port=5600 caps='application/x-rtp, media=(string)video, clock-rate=(int)90000, encoding-name=(string)H264' ! rtph264depay ! 'video/x-h264,stream-format=byte-stream' ! fdsink | fpv_video"`

### RTP MMAL
`ExecStart=/bin/sh -c "gst-launch-1.0 udpsrc port=5600 caps='application/x-rtp, media=(string)video, clock-rate=(int)90000, encoding-name=(string)H264' ! rtph264depay ! h264parse disable-passthrough=true config-interval=1 ! queue ! mmalh264dec ! mmalvideosink sync=false"`


**d) Because I'm not familiar with the gst tool. Is the following command saying that fpv-camera is using an RTSP camera as a video source?**

### RTSP legacy
#ExecStart=/bin/sh -c "gst-launch-1.0 rtspsrc latency=0 location="rtsp://127.0.0.1:8554/wfb" ! rtph264depay ! 'video/x-h264,stream-format=byte-stream' ! fdsink | fpv_video"

### RTSP MMAL
`ExecStart=/bin/sh -c "gst-launch-1.0 rtspsrc latency=0 location="rtsp://127.0.0.1:8554/wfb" ! rtph264depay ! h264parse disable-passthrough=true config-interval=1 ! queue ! mmalh264dec ! mmalvideosink sync=false"`

**Reply by svpcom** [reply]
Both of them use h264 hw codecs, but mmal has better integration with gstreamer. I prefer to use MMAL, but some crap cameras supply video streams that mmal unable to handle right, so I leave option for legacy pipeline

**Reply by lida2003** [reply]
what about rtsp? How can i turn rtsp server on?

> **d) Because I'm not familiar with the gst tool. Is the following command saying that fpv-camera is using an RTSP camera as a video source?**
> 
> ### RTSP legacy
> #ExecStart=/bin/sh -c "gst-launch-1.0 rtspsrc latency=0 location="rtsp://127.0.0.1:8554/wfb" ! rtph264depay ! 'video/x-h264,stream-format=byte-stream' ! fdsink | fpv_video"
> 
> ### RTSP MMAL
> `ExecStart=/bin/sh -c "gst-launch-1.0 rtspsrc latency=0 location="rtsp://127.0.0.1:8554/wfb" ! rtph264depay ! h264parse disable-passthrough=true config-interval=1 ! queue ! mmalh264dec ! mmalvideosink sync=false"`
>

**Reply by svpcom** [reply]
`sudo systemctl start rtsp`

**Reply by svpcom** [reply]
In this example RTSP used on **GS** side to distribute the same video steam to multiple viewers

**Reply by svpcom** [reply]
For example you can see video on RPI hdmi output and in QGC on the laptop simulteneously

**Reply by lida2003** [reply]
> For example you can see video on RPI hdmi output and in QGC on the laptop simulteneously

OK, I got it. How about the latency when using RTP  --> RTSP --> FPV_video, instead of RTP --> FPV_video?

**Reply by lida2003** [reply]
> `sudo systemctl start rtsp`

Thanks, I got the code [test-video.diff.zip](https://github.com/svpcom/wfb-ng/files/11024439/test-video.diff.zip), which is much more clear of the steaming path.

**Reply by svpcom** [reply]
RTSP is just signalling protocol. All data are transferred via RTP as before, but now RTP streams are dynamically configured by RTSP server. So latency via RTSP is the same as via RTP

**Reply by svpcom** [reply]
Yes, I leave patched gst rstp server sources for reference.
